File: frontend/lib/hexagrams.ts
// Единый источник истины для гексаграмм (номер, комбинация, название, стадия).
// Потребители: admin/strategies/[combination], report/[id], AdminNav.
// Имена гексаграмм меняем ТОЛЬКО здесь.
export type Hexagram = { n: number; combo: string; name: string };

export const HEXAGRAM_DATA: Hexagram[] = [
  { n:  1, combo: 'AAAAAA', name: 'Действие' },
  { n:  2, combo: 'BBBBBB', name: 'Реакция' },
  { n:  3, combo: 'ABBBAB', name: 'Появление' },
  { n:  4, combo: 'BABBBA', name: 'Формализация' },
  { n:  5, combo: 'AAABAB', name: 'Бдительность' },
  { n:  6, combo: 'BABAAA', name: 'Раздор' },
  { n:  7, combo: 'BABBBB', name: 'Управление' },
  { n:  8, combo: 'BBBBAB', name: 'Сближение' },
  { n:  9, combo: 'AAABAA', name: 'Развитие' },
  { n: 10, combo: 'AABAAA', name: 'Последовательность' },
  { n: 11, combo: 'AAABBB', name: 'Достижение' },
  { n: 12, combo: 'BBBAAA', name: 'Препятствие' },
  { n: 13, combo: 'ABAAAA', name: 'Осознанность' },
  { n: 14, combo: 'AAAABA', name: 'Процветание' },
  { n: 15, combo: 'BBABBB', name: 'Смирение' },
  { n: 16, combo: 'BBBABB', name: 'Радость' },
  { n: 17, combo: 'ABBAAB', name: 'Соответствие' },
  { n: 18, combo: 'BAABBA', name: 'Диссонанс' },
  { n: 19, combo: 'AABBBB', name: 'Подход' },
  { n: 20, combo: 'BBBBAA', name: 'Наблюдать' },
  { n: 21, combo: 'ABBABA', name: 'Устранять' },
  { n: 22, combo: 'ABABBA', name: 'Изящество' },
  { n: 23, combo: 'BBBBBA', name: 'Разрушение' },
  { n: 24, combo: 'ABBBBB', name: 'Возрождение' },
  { n: 25, combo: 'ABBAAA', name: 'Естественность' },
  { n: 26, combo: 'AAABBA', name: 'Накопление' },
  { n: 27, combo: 'ABBBBA', name: 'Умеренность' },
  { n: 28, combo: 'BAAAAB', name: 'Избыток' },
  { n: 29, combo: 'BABBAB', name: 'Решимость' },
  { n: 30, combo: 'ABAABA', name: 'Великолепие' },
  { n: 31, combo: 'BBAAAB', name: 'Влияние' },
  { n: 32, combo: 'BAAABB', name: 'Выносливость' },
  { n: 33, combo: 'BBAAAA', name: 'Благоразумие' },
  { n: 34, combo: 'AAAABB', name: 'Сила' },
  { n: 35, combo: 'BBBABA', name: 'Благоприятный' },
  { n: 36, combo: 'ABABBB', name: 'Неблагоприятный' },
  { n: 37, combo: 'ABABAA', name: 'Гармония' },
  { n: 38, combo: 'AABABA', name: 'Полярность' },
  { n: 39, combo: 'BBABAB', name: 'Трудность' },
  { n: 40, combo: 'BABABB', name: 'Избавление' },
  { n: 41, combo: 'AABBBA', name: 'Убыток' },
  { n: 42, combo: 'ABBBAA', name: 'Прибыль' },
  { n: 43, combo: 'AAAAAB', name: 'Прорыв' },
  { n: 44, combo: 'BAAAAA', name: 'Встреча' },
  { n: 45, combo: 'BBBAAB', name: 'Объединение' },
  { n: 46, combo: 'BAABBB', name: 'Самоотдача' },
  { n: 47, combo: 'BABAAB', name: 'Понимание' },
  { n: 48, combo: 'BAABAB', name: 'Глубина' },
  { n: 49, combo: 'ABAAAB', name: 'Реформа' },
  { n: 50, combo: 'BAAABA', name: 'Ценности' },
  { n: 51, combo: 'ABBABB', name: 'Смелость' },
  { n: 52, combo: 'BBABBA', name: 'Сосредоточенность' },
  { n: 53, combo: 'BBABAA', name: 'Готовность' },
  { n: 54, combo: 'AABABB', name: 'Амбиции' },
  { n: 55, combo: 'ABAABB', name: 'Изобилие' },
  { n: 56, combo: 'BBAABA', name: 'Стимулирование' },
  { n: 57, combo: 'BAABAA', name: 'Интуиция' },
  { n: 58, combo: 'AABAAB', name: 'Бодрость' },
  { n: 59, combo: 'BABBAA', name: 'Установление связей' },
  { n: 60, combo: 'AABBAB', name: 'Реализм' },
  { n: 61, combo: 'AABBAA', name: 'Внутренняя правда' },
  { n: 62, combo: 'BBAABB', name: 'Точность' },
  { n: 63, combo: 'ABABAB', name: 'Завершение' },
  { n: 64, combo: 'BABABA', name: 'Незавершённость' },
];

export const HEXAGRAM_MAP: Record<string, Hexagram> = {};
for (const h of HEXAGRAM_DATA) HEXAGRAM_MAP[h.combo] = h;

export const HEX_TUPLE: Record<string, [number, string]> = {};
for (const h of HEXAGRAM_DATA) HEX_TUPLE[h.combo] = [h.n, h.name];

export function comboToHex(combo: string): string {
  const e = HEXAGRAM_MAP[combo];
  return e ? String.fromCodePoint(0x4DC0 + e.n - 1) : '?';
}

File: backend/app/hexagrams.py
# -*- coding: utf-8 -*-
# Единый источник истины для гексаграмм: (номер, название, комбинация).
# Потребитель: pdf.py. Имена гексаграмм меняем ТОЛЬКО здесь.

HEXAGRAM_LIST: list[tuple[int, str, str]] = [
    (1,  "Действие",              "AAAAAA"),
    (2,  "Реакция",               "BBBBBB"),
    (3,  "Появление",             "ABBBAB"),
    (4,  "Формализация",          "BABBBA"),
    (5,  "Бдительность",          "AAABAB"),
    (6,  "Раздор",                "BABAAA"),
    (7,  "Управление",            "BABBBB"),
    (8,  "Сближение",             "BBBBAB"),
    (9,  "Развитие",              "AAABAA"),
    (10, "Последовательность",    "AABAAA"),
    (11, "Достижение",            "AAABBB"),
    (12, "Препятствие",           "BBBAAA"),
    (13, "Осознанность",          "ABAAAA"),
    (14, "Процветание",           "AAAABA"),
    (15, "Смирение",              "BBABBB"),
    (16, "Радость",               "BBBABB"),
    (17, "Соответствие",          "ABBAAB"),
    (18, "Диссонанс",             "BAABBA"),
    (19, "Подход",                "AABBBB"),
    (20, "Наблюдать",             "BBBBAA"),
    (21, "Устранять",             "ABBABA"),
    (22, "Изящество",             "ABABBA"),
    (23, "Разрушение",            "BBBBBA"),
    (24, "Возрождение",           "ABBBBB"),
    (25, "Естественность",        "ABBAAA"),
    (26, "Накопление",            "AAABBA"),
    (27, "Умеренность",           "ABBBBA"),
    (28, "Избыток",               "BAAAAB"),
    (29, "Решимость",             "BABBAB"),
    (30, "Великолепие",           "ABAABA"),
    (31, "Влияние",               "BBAAAB"),
    (32, "Выносливость",          "BAAABB"),
    (33, "Благоразумие",          "BBAAAA"),
    (34, "Сила",                  "AAAABB"),
    (35, "Благоприятный",         "BBBABA"),
    (36, "Неблагоприятный",       "ABABBB"),
    (37, "Гармония",              "ABABAA"),
    (38, "Полярность",            "AABABA"),
    (39, "Трудность",             "BBABAB"),
    (40, "Избавление",            "BABABB"),
    (41, "Убыток",                "AABBBA"),
    (42, "Прибыль",               "ABBBAA"),
    (43, "Прорыв",                "AAAAAB"),
    (44, "Встреча",               "BAAAAA"),
    (45, "Объединение",           "BBBAAB"),
    (46, "Самоотдача",            "BAABBB"),
    (47, "Понимание",             "BABAAB"),
    (48, "Глубина",               "BAABAB"),
    (49, "Реформа",               "ABAAAB"),
    (50, "Ценности",              "BAAABA"),
    (51, "Смелость",              "ABBABB"),
    (52, "Сосредоточенность",     "BBABBA"),
    (53, "Готовность",            "BBABAA"),
    (54, "Амбиции",               "AABABB"),
    (55, "Изобилие",              "ABAABB"),
    (56, "Стимулирование",        "BBAABA"),
    (57, "Интуиция",              "BAABAA"),
    (58, "Бодрость",              "AABAAB"),
    (59, "Установление связей",   "BABBAA"),
    (60, "Реализм",               "AABBAB"),
    (61, "Внутренняя правда",     "AABBAA"),
    (62, "Точность",              "BBAABB"),
    (63, "Завершение",            "ABABAB"),
    (64, "Незавершённость",       "BABABA"),
]


_HEXAGRAM_LIST = HEXAGRAM_LIST

# combination → (number, name)
_HEXAGRAM_BY_COMBO: dict[str, tuple[int, str]] = {
    combo: (num, name) for num, name, combo in _HEXAGRAM_LIST
}

# number → name
_HEXAGRAM_BY_NUM: dict[int, str] = {
    num: name for num, name, _ in _HEXAGRAM_LIST
}

# number → combination
_COMBO_BY_NUM: dict[int, str] = {num: combo for num, name, combo in _HEXAGRAM_LIST}

# Таблица соответствия: номер текущей гексаграммы → номер целевой
_TARGET_HEXAGRAM: dict[int, int] = {
     1:  9,  2: 62,  3: 49,  4:  7,  5: 63,  6:  6,  7: 62,  8: 23,
     9: 37, 10: 25, 11: 36, 12:  9, 13: 37, 14: 26, 15: 11, 16: 54,
    17: 63, 18: 64, 19: 34, 20: 33, 21: 64, 22: 18, 23: 56, 24: 19,
    25: 37, 26: 22, 27:  4, 28: 44, 29:  3, 30: 22, 31: 43, 32: 44,
    33:  1, 34:  1, 35: 64, 36: 37, 37: 63, 38: 21, 39:  5, 40: 46,
    41: 27, 42:  3, 43:  5, 44: 33, 45: 58, 46: 57, 47: 44, 48: 47,
    49: 63, 50: 18, 51: 25, 52: 18, 53: 39, 54: 11, 55: 36, 56: 14,
    57: 44, 58:  5, 59: 44, 60: 43, 61: 42, 62: 33, 63: 17, 64: 40,
}


def get_target_hexagram_info(combination: str) -> tuple[int, str, str] | None:
    """
    По комбинации (напр. ABABBA) возвращает (номер, название, символ)
    целевой гексаграммы. Символ — Unicode U+4DC0 + num - 1.
    Возвращает None, если комбинация не найдена.
    """
    entry = _HEXAGRAM_BY_COMBO.get(combination)
    if not entry:
        return None
    current_num, _ = entry
    target_num = _TARGET_HEXAGRAM.get(current_num)
    if not target_num:
        return None
    target_name = _HEXAGRAM_BY_NUM.get(target_num, "")
    target_symbol = chr(0x4DC0 + target_num - 1)
    return target_num, target_name, target_symbol


def hexagram_symbol(num: int | None) -> str:
    """Unicode-символ гексаграммы по номеру: U+4DC0 + num - 1."""
    return chr(0x4DC0 + num - 1) if num else ''
